File: airbyte-cdk/java/airbyte-cdk/db-sources/src/main/java/io/airbyte/cdk/integrations/source/relationaldb/RelationalDbReadUtil.java
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.cdk.integrations.source.relationaldb;

import com.google.common.collect.Sets;
import io.airbyte.commons.json.Jsons;
import io.airbyte.protocol.models.v0.AirbyteStreamNameNamespacePair;
import io.airbyte.protocol.models.v0.ConfiguredAirbyteCatalog;
import io.airbyte.protocol.models.v0.ConfiguredAirbyteStream;
import io.airbyte.protocol.models.v0.SyncMode;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

public class RelationalDbReadUtil {

  public static List<ConfiguredAirbyteStream> identifyStreamsToSnapshot(final ConfiguredAirbyteCatalog catalog,
                                                                        final Set<AirbyteStreamNameNamespacePair> alreadySyncedStreams) {
    final Set<AirbyteStreamNameNamespacePair> allStreams = AirbyteStreamNameNamespacePair.fromConfiguredCatalog(catalog);
    final Set<AirbyteStreamNameNamespacePair> newlyAddedStreams = new HashSet<>(Sets.difference(allStreams, alreadySyncedStreams));
    return catalog.getStreams().stream()
        .filter(c -> c.getSyncMode() == SyncMode.INCREMENTAL)
        .filter(stream -> newlyAddedStreams.contains(AirbyteStreamNameNamespacePair.fromAirbyteStream(stream.getStream())))
        .map(Jsons::clone)
        .collect(Collectors.toList());
  }

  public static List<ConfiguredAirbyteStream> identifyStreamsForCursorBased(final ConfiguredAirbyteCatalog catalog,
                                                                            final List<ConfiguredAirbyteStream> streamsForInitialLoad) {

    final Set<AirbyteStreamNameNamespacePair> initialLoadStreamsNamespacePairs =
        streamsForInitialLoad.stream().map(stream -> AirbyteStreamNameNamespacePair.fromAirbyteStream(stream.getStream()))
            .collect(
                Collectors.toSet());
    return catalog.getStreams().stream()
        .filter(c -> c.getSyncMode() == SyncMode.INCREMENTAL)
        .filter(stream -> !initialLoadStreamsNamespacePairs.contains(AirbyteStreamNameNamespacePair.fromAirbyteStream(stream.getStream())))
        .map(Jsons::clone)
        .collect(Collectors.toList());
  }

  public static AirbyteStreamNameNamespacePair convertNameNamespacePairFromV0(final io.airbyte.protocol.models.AirbyteStreamNameNamespacePair v1NameNamespacePair) {
    return new AirbyteStreamNameNamespacePair(v1NameNamespacePair.getName(), v1NameNamespacePair.getNamespace());
  }

}

File: airbyte-cdk/java/airbyte-cdk/db-sources/src/main/kotlin/io/airbyte/cdk/integrations/source/relationaldb/RelationalDbReadUtil.kt
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */
package io.airbyte.cdk.integrations.source.relationaldb

import com.google.common.collect.Sets
import io.airbyte.commons.json.Jsons
import io.airbyte.protocol.models.v0.AirbyteStreamNameNamespacePair
import io.airbyte.protocol.models.v0.ConfiguredAirbyteCatalog
import io.airbyte.protocol.models.v0.ConfiguredAirbyteStream
import io.airbyte.protocol.models.v0.SyncMode

object RelationalDbReadUtil {
    fun identifyStreamsToSnapshot(
        catalog: ConfiguredAirbyteCatalog,
        alreadySyncedStreams: Set<AirbyteStreamNameNamespacePair>
    ): List<ConfiguredAirbyteStream> {
        val allStreams = AirbyteStreamNameNamespacePair.fromConfiguredCatalog(catalog)
        val newlyAddedStreams: Set<AirbyteStreamNameNamespacePair> =
            HashSet(Sets.difference(allStreams, alreadySyncedStreams))
        return catalog.streams
            .filter { c: ConfiguredAirbyteStream -> c.syncMode == SyncMode.INCREMENTAL }
            .filter { stream: ConfiguredAirbyteStream ->
                newlyAddedStreams.contains(
                    AirbyteStreamNameNamespacePair.fromAirbyteStream(stream.stream)
                )
            }
            .map { `object`: ConfiguredAirbyteStream -> Jsons.clone(`object`) }
    }

    @JvmStatic
    fun identifyStreamsForCursorBased(
        catalog: ConfiguredAirbyteCatalog,
        streamsForInitialLoad: List<ConfiguredAirbyteStream>
    ): List<ConfiguredAirbyteStream> {
        val initialLoadStreamsNamespacePairs =
            streamsForInitialLoad
                .map { stream: ConfiguredAirbyteStream ->
                    AirbyteStreamNameNamespacePair.fromAirbyteStream(stream.stream)
                }
                .toSet()
        return catalog.streams
            .filter { c: ConfiguredAirbyteStream -> c.syncMode == SyncMode.INCREMENTAL }
            .filter { stream: ConfiguredAirbyteStream ->
                !initialLoadStreamsNamespacePairs.contains(
                    AirbyteStreamNameNamespacePair.fromAirbyteStream(stream.stream)
                )
            }
            .map { `object`: ConfiguredAirbyteStream -> Jsons.clone(`object`) }
    }

    @JvmStatic
    fun convertNameNamespacePairFromV0(
        v1NameNamespacePair: io.airbyte.protocol.models.AirbyteStreamNameNamespacePair
    ): AirbyteStreamNameNamespacePair {
        return AirbyteStreamNameNamespacePair(
            v1NameNamespacePair.name,
            v1NameNamespacePair.namespace
        )
    }
}
